File: haskell/src/DCF/Transport/FEC.hs
-- SPDX-License-Identifier: LGPL-3.0-only

-- | DCF forward-error-correction adapter — systematic Reed-Solomon over GF(2^8),
-- byte-identical to feclab_core.py, codec/demod_fec.h, codec/src/fec.rs, go/dcf/fec.go,
-- JS/nodejs/src/fec.js, cpp/include/dcf/fec.hpp, perl/lib/DCF/FEC.pm, lua/dcf_fec.lua,
-- and lisp/src/fec.lisp (pinned by Documentation/fec_vectors.json). FEC wraps a 17-byte
-- DeModFrame so a lossy medium (RF/SDR, acoustic) can CORRECT it, not just detect.
--
-- Field GF(2^8), prim 0x11D, generator 2, fcr 0. Systematic: codeword =
-- message ++ parity. Default 2t=16 parity -> corrects 8 byte-errors. The message
-- layer chunks + interleaves any-length payloads behind a self-protecting header.
module DCF.Transport.FEC
  ( rsEncode
  , rsDecode
  , interleave
  , deinterleave
  , encodeMessage
  , decodeMessage
  , rsDefaultParity
  ) where

import           Control.Monad       (forM, forM_, when)
import           Control.Monad.ST    (ST, runST)
import           Data.Array          (Array, listArray, (!))
import           Data.Array.ST       (STUArray, newArray, readArray, writeArray)
import           Data.Bits           (shiftL, shiftR, xor, (.&.), (.|.))
import           Data.List           (foldl')
import           Data.Word           (Word8)

rsDefaultParity :: Int
rsDefaultParity = 16

gfPrim :: Int
gfPrim = 0x11D

expList :: [Int]
expList = take 255 (iterate step 1)
  where step x = let y = shiftL x 1 in if y .&. 0x100 /= 0 then y `xor` gfPrim else y

gfExp :: Array Int Int
gfExp = listArray (0, 511) (expList ++ [ expList !! (i `mod` 255) | i <- [255 .. 511] ])

gfLog :: Array Int Int
gfLog = listArray (0, 255) [ logOf b | b <- [0 .. 255] ]
  where pairs = zip expList [0 ..]
        logOf 0 = 0
        logOf b = maybe 0 id (lookup b pairs)

gmul :: Int -> Int -> Int
gmul a b | a == 0 || b == 0 = 0
         | otherwise = gfExp ! (gfLog ! a + gfLog ! b)
gdiv :: Int -> Int -> Int
gdiv a b | a == 0 = 0
         | otherwise = gfExp ! ((gfLog ! a - gfLog ! b) `mod` 255)
gpow :: Int -> Int -> Int
gpow a p = gfExp ! ((gfLog ! a * p) `mod` 255)
ginv :: Int -> Int
ginv a = gfExp ! (255 - gfLog ! a)

-- Polynomials as [Int], high-order coefficient first.
pmul :: [Int] -> [Int] -> [Int]
pmul p q = [ foldl' xor 0 [ gmul (p !! i) (q !! (k - i))
                          | i <- [max 0 (k - (length q - 1)) .. min k (length p - 1)] ]
           | k <- [0 .. length p + length q - 2] ]
padd :: [Int] -> [Int] -> [Int]
padd p q = zipWith xor (pad p) (pad q)
  where n = max (length p) (length q)
        pad xs = replicate (n - length xs) 0 ++ xs
pscale :: [Int] -> Int -> [Int]
pscale p s = map (`gmul` s) p
peval :: [Int] -> Int -> Int
peval p x = foldl' (\y c -> gmul y x `xor` c) (head p) (tail p)

genpoly :: Int -> [Int]
genpoly np = foldl' (\g i -> pmul g [1, gpow 2 i]) [1] [0 .. np - 1]

-- ── encode (systematic, via the reference's polynomial division) ──────────────
rsEncodeI :: [Int] -> Int -> [Int]
rsEncodeI msg np = msg ++ runST go
  where
    m   = length msg
    gen = genpoly np
    go :: ST s [Int]
    go = do
      arr <- newArray (0, m + np - 1) 0 :: ST s (STUArray s Int Int)
      forM_ (zip [0 ..] msg) $ \(i, b) -> writeArray arr i b
      forM_ [0 .. m - 1] $ \i -> do
        coef <- readArray arr i
        when (coef /= 0) $
          forM_ [1 .. np] $ \j -> do
            v <- readArray arr (i + j)
            writeArray arr (i + j) (v `xor` gmul (gen !! j) coef)
      forM [m .. m + np - 1] (readArray arr)

rsEncode :: [Word8] -> Int -> [Word8]
rsEncode msg np = map fromIntegral (rsEncodeI (map fromIntegral msg) np)

-- ── decode (Berlekamp–Massey -> Chien -> Forney) ──────────────────────────────
syndromes :: [Int] -> Int -> [Int]
syndromes cw np = 0 : [ peval cw (gpow 2 i) | i <- [0 .. np - 1] ]

errLocator :: [Int] -> Int -> [Int]
errLocator synd np = dropWhile (== 0) (go 0 [1] [1])
  where
    go i el ol
      | i >= np = el
      | otherwise =
          let delta = foldl' xor (synd !! (i + 1))
                        [ gmul (el !! (length el - 1 - j)) (synd !! (i + 1 - j)) | j <- [1 .. length el - 1] ]
              ol' = ol ++ [0]
          in if delta == 0
               then go (i + 1) el ol'
               else let (el2, ol2) = if length ol' > length el
                                       then (pscale ol' delta, pscale el (ginv delta))
                                       else (el, ol')
                    in go (i + 1) (padd el2 (pscale ol2 delta)) ol2

rsDecodeI :: [Int] -> Int -> Int -> Either String ([Int], Int)
rsDecodeI codeword np msglen =
  let synd = syndromes codeword np
  in if all (== 0) synd
       then Right (take msglen codeword, 0)
       else
         let el  = errLocator synd np
             rev = reverse el
             errs = length rev - 1
             n   = length codeword
             pos = [ n - 1 - i | i <- [0 .. n - 1], peval rev (gpow 2 i) == 0 ]
         in if length el - 1 > np `div` 2 then Left "too many errors"
            else if length pos /= errs then Left "error location failed"
            else
              let coefPos = map (\p -> n - 1 - p) pos
                  eloc = foldl' (\e cp -> pmul e [gpow 2 cp, 1]) [1] coefPos
                  xs   = map (\cp -> gpow 2 cp) coefPos
                  prod = pmul (reverse synd) eloc
                  rem' = drop (length prod - length eloc) prod
                  fixed = runST $ do
                    arr <- newArray (0, n - 1) 0 :: ST s (STUArray s Int Int)
                    forM_ (zip [0 ..] codeword) $ \(i, b) -> writeArray arr i b
                    forM_ (zip [0 ..] xs) $ \(i, xi) -> do
                      let xinv  = ginv xi
                          denom = foldl' (\d (j, xj) -> if j /= i then gmul d (1 `xor` gmul xinv xj) else d)
                                         1 (zip [0 ..] xs)
                          numer = gmul (peval rem' xinv) (gpow xi 1)
                      v <- readArray arr (pos !! i)
                      writeArray arr (pos !! i) (v `xor` gdiv numer denom)
                    forM [0 .. n - 1] (readArray arr)
              in if all (== 0) (syndromes fixed np)
                   then Right (take msglen fixed, length pos)
                   else Left "residual syndrome"

rsDecode :: [Word8] -> Int -> Maybe Int -> Either String ([Word8], Int)
rsDecode codeword np mlen =
  let ci = map fromIntegral codeword
      ml = maybe (length ci - np) id mlen
  in case rsDecodeI ci np ml of
       Left e        -> Left e
       Right (m, nc) -> Right (map fromIntegral m, nc)

-- ── interleaver + multi-codeword messages ─────────────────────────────────────
interleave :: [[Word8]] -> [Word8]
interleave [] = []
interleave cws = [ (cws !! r) !! c | c <- [0 .. n - 1], r <- [0 .. d - 1] ]
  where d = length cws; n = length (head cws)

deinterleave :: [Word8] -> Int -> Int -> [[Word8]]
deinterleave stream depth n =
  [ [ stream !! (c * depth + r) | c <- [0 .. n - 1] ] | r <- [0 .. depth - 1] ]

hdrParity :: Int
hdrParity = 16
hdrLen :: Int
hdrLen = 21

chunking :: Int -> Int -> (Int, Int)
chunking l np
  | l == 0    = (1, 1)
  | otherwise = let nchunks = (l + maxk - 1) `div` maxk
                    k       = (l + nchunks - 1) `div` nchunks
                in (nchunks, k)
  where maxk = 255 - np

encodeMessage :: [Word8] -> Int -> [Word8]
encodeMessage msg np =
  let l = length msg
      (nchunks, k) = chunking l np
      hdr = [ fromIntegral (shiftR l 24 .&. 255), fromIntegral (shiftR l 16 .&. 255)
            , fromIntegral (shiftR l 8 .&. 255),  fromIntegral (l .&. 255)
            , fromIntegral np ]
      cws = [ rsEncode (block c) np | c <- [0 .. nchunks - 1] ]
      block c = take k (drop (c * k) msg ++ replicate k 0)
  in rsEncode hdr hdrParity ++ interleave cws

decodeMessage :: [Word8] -> Either String ([Word8], Int)
decodeMessage blob
  | length blob < hdrLen = Left "short blob"
  | otherwise =
      case rsDecode (take hdrLen blob) hdrParity (Just 5) of
        Left e        -> Left e
        Right (h, _)  ->
          let l  = (fromIntegral (h !! 0) `shiftL` 24) .|. (fromIntegral (h !! 1) `shiftL` 16)
                   .|. (fromIntegral (h !! 2) `shiftL` 8) .|. fromIntegral (h !! 3) :: Int
              np = fromIntegral (h !! 4)
              (nchunks, k) = chunking l np
              cwlen = k + np
              body  = drop hdrLen blob
          in if length body /= nchunks * cwlen then Left "blob length mismatch"
             else
               let results = map (\cw -> rsDecode cw np (Just k)) (deinterleave body nchunks cwlen)
               in case sequence results of
                    Left e   -> Left e
                    Right rs -> Right (take l (concatMap fst rs), sum (map snd rs))

File: java/com/demod/dcf/FEC.java
// SPDX-License-Identifier: LGPL-3.0-only
package com.demod.dcf;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * DCF forward-error-correction adapter — systematic Reed-Solomon over GF(2^8),
 * byte-identical to feclab_core.py, codec/demod_fec.h, codec/src/fec.rs, go/dcf/fec.go,
 * JS/nodejs/src/fec.js, cpp/include/dcf/fec.hpp, perl/lib/DCF/FEC.pm, lua/dcf_fec.lua,
 * lisp/src/fec.lisp, and DCF.Transport.FEC (pinned by Documentation/fec_vectors.json).
 * FEC wraps a 17-byte DeModFrame so a lossy medium can CORRECT it, not just detect.
 *
 * <p>Field GF(2^8), prim 0x11D, generator 2, fcr 0. Systematic: codeword =
 * message ++ parity. Default 2t=16 parity -&gt; corrects 8 byte-errors. The message
 * layer chunks + interleaves any-length payloads behind a self-protecting header.
 */
public final class FEC {

    public static final int GF_PRIM = 0x11D;
    public static final int GF_GEN = 2;
    public static final int FCR = 0;
    public static final int RS_DEFAULT_PARITY = 16;
    public static final int HDR_PARITY = 16;
    public static final int HDR_LEN = 21;

    private static final int[] EXP = new int[512];
    private static final int[] LOG = new int[256];
    static {
        int x = 1;
        for (int i = 0; i < 255; i++) {
            EXP[i] = x;
            LOG[x] = i;
            x <<= 1;
            if ((x & 0x100) != 0) {
                x ^= GF_PRIM;
            }
        }
        for (int i = 255; i < 512; i++) {
            EXP[i] = EXP[i - 255];
        }
    }

    private FEC() {}

    static int gmul(int a, int b) { return (a == 0 || b == 0) ? 0 : EXP[LOG[a] + LOG[b]]; }
    static int gdiv(int a, int b) { return a == 0 ? 0 : EXP[(LOG[a] + 255 - LOG[b]) % 255]; }
    static int gpow(int a, int p) { return EXP[((LOG[a] * p) % 255 + 255) % 255]; }
    static int ginv(int a) { return EXP[255 - LOG[a]]; }

    static int[] pmul(int[] p, int[] q) {
        int[] r = new int[p.length + q.length - 1];
        for (int j = 0; j < q.length; j++) {
            for (int i = 0; i < p.length; i++) {
                r[i + j] ^= gmul(p[i], q[j]);
            }
        }
        return r;
    }
    static int[] padd(int[] p, int[] q) {
        int n = Math.max(p.length, q.length);
        int[] r = new int[n];
        for (int i = 0; i < p.length; i++) { r[i + n - p.length] = p[i]; }
        for (int i = 0; i < q.length; i++) { r[i + n - q.length] ^= q[i]; }
        return r;
    }
    static int[] pscale(int[] p, int s) {
        int[] r = new int[p.length];
        for (int i = 0; i < p.length; i++) { r[i] = gmul(p[i], s); }
        return r;
    }
    static int peval(int[] p, int x) {
        int y = p[0];
        for (int i = 1; i < p.length; i++) { y = gmul(y, x) ^ p[i]; }
        return y;
    }
    static int[] genpoly(int np) {
        int[] g = {1};
        for (int i = 0; i < np; i++) { g = pmul(g, new int[] {1, gpow(GF_GEN, FCR + i)}); }
        return g;
    }

    /** message bytes -&gt; codeword bytes (message ++ nparity parity), systematic. */
    public static byte[] rsEncode(byte[] msg, int np) {
        int m = msg.length;
        int[] gen = genpoly(np);
        int[] out = new int[m + np];
        for (int i = 0; i < m; i++) { out[i] = msg[i] & 0xFF; }
        for (int i = 0; i < m; i++) {
            int coef = out[i];
            if (coef != 0) {
                for (int j = 1; j < gen.length; j++) { out[i + j] ^= gmul(gen[j], coef); }
            }
        }
        byte[] code = new byte[m + np];
        for (int i = 0; i < m; i++) { code[i] = msg[i]; }
        for (int i = 0; i < np; i++) { code[m + i] = (byte) out[m + i]; }
        return code;
    }

    /** Decode result: the recovered message and the number of corrected bytes. */
    public static final class Result {
        public final byte[] msg;
        public final int corrected;
        Result(byte[] m, int c) { msg = m; corrected = c; }
    }

    static int[] syndromes(int[] cw, int np) {
        int[] s = new int[np + 1];
        for (int i = 0; i < np; i++) { s[i + 1] = peval(cw, gpow(GF_GEN, FCR + i)); }
        return s;
    }
    static int[] errLocator(int[] synd, int np) {
        int[] el = {1};
        int[] ol = {1};
        for (int i = 0; i < np; i++) {
            int delta = synd[i + 1];
            for (int j = 1; j < el.length; j++) { delta ^= gmul(el[el.length - 1 - j], synd[i + 1 - j]); }
            ol = Arrays.copyOf(ol, ol.length + 1);
            if (delta != 0) {
                if (ol.length > el.length) {
                    int[] nl = pscale(ol, delta);
                    ol = pscale(el, ginv(delta));
                    el = nl;
                }
                el = padd(el, pscale(ol, delta));
            }
        }
        int s = 0;
        while (s < el.length && el[s] == 0) { s++; }
        return Arrays.copyOfRange(el, s, el.length);
    }
    static int[] reverse(int[] a) {
        int[] r = new int[a.length];
        for (int i = 0; i < a.length; i++) { r[i] = a[a.length - 1 - i]; }
        return r;
    }

    /** codeword bytes -&gt; Result. msglen &lt; 0 means codeword.length - nparity. */
    public static Result rsDecode(byte[] codeword, int np, int msglen) {
        int n = codeword.length;
        int[] cw = new int[n];
        for (int i = 0; i < n; i++) { cw[i] = codeword[i] & 0xFF; }
        if (msglen < 0) { msglen = n - np; }
        int[] synd = syndromes(cw, np);
        boolean zero = true;
        for (int s : synd) { if (s != 0) { zero = false; break; } }
        if (zero) { return new Result(slice(cw, msglen), 0); }
        int[] el = errLocator(synd, np);
        if (el.length - 1 > np / 2) { throw new RuntimeException("fec: too many errors"); }
        int[] rev = reverse(el);
        int errs = rev.length - 1;
        List<Integer> pos = new ArrayList<>();
        for (int i = 0; i < n; i++) { if (peval(rev, gpow(GF_GEN, i)) == 0) { pos.add(n - 1 - i); } }
        if (pos.size() != errs) { throw new RuntimeException("fec: error location failed"); }
        int[] eloc = {1};
        int[] xs = new int[pos.size()];
        for (int i = 0; i < pos.size(); i++) {
            int cp = n - 1 - pos.get(i);
            eloc = pmul(eloc, new int[] {gpow(GF_GEN, cp), 1});
            xs[i] = gpow(GF_GEN, cp);
        }
        int[] prod = pmul(reverse(synd), eloc);
        int[] rem = Arrays.copyOfRange(prod, prod.length - eloc.length, prod.length);
        for (int i = 0; i < xs.length; i++) {
            int xi = xs[i];
            int xiInv = ginv(xi);
            int denom = 1;
            for (int j = 0; j < xs.length; j++) { if (j != i) { denom = gmul(denom, 1 ^ gmul(xiInv, xs[j])); } }
            int numer = gmul(peval(rem, xiInv), gpow(xi, 1 - FCR));
            if (denom == 0) { throw new RuntimeException("fec: forney denom zero"); }
            cw[pos.get(i)] ^= gdiv(numer, denom);
        }
        for (int s : syndromes(cw, np)) { if (s != 0) { throw new RuntimeException("fec: residual syndrome"); } }
        return new Result(slice(cw, msglen), pos.size());
    }
    static byte[] slice(int[] cw, int len) {
        byte[] b = new byte[len];
        for (int i = 0; i < len; i++) { b[i] = (byte) cw[i]; }
        return b;
    }

    public static byte[] interleave(byte[][] cws) {
        if (cws.length == 0) { return new byte[0]; }
        int d = cws.length;
        int n = cws[0].length;
        byte[] out = new byte[d * n];
        for (int r = 0; r < d; r++) { for (int c = 0; c < n; c++) { out[c * d + r] = cws[r][c]; } }
        return out;
    }
    public static byte[][] deinterleave(byte[] stream, int depth, int n) {
        byte[][] cws = new byte[depth][n];
        for (int c = 0; c < n; c++) { for (int r = 0; r < depth; r++) { cws[r][c] = stream[c * depth + r]; } }
        return cws;
    }
    static int[] chunking(int l, int np) {
        int maxk = 255 - np;
        int nchunks = l == 0 ? 1 : (l + maxk - 1) / maxk;
        int k = l == 0 ? 1 : (l + nchunks - 1) / nchunks;
        return new int[] {nchunks, k};
    }

    public static byte[] encodeMessage(byte[] msg, int np) {
        int l = msg.length;
        int[] ck = chunking(l, np);
        int nchunks = ck[0];
        int k = ck[1];
        byte[] hdr = {(byte) (l >> 24), (byte) (l >> 16), (byte) (l >> 8), (byte) l, (byte) np};
        byte[] out = rsEncode(hdr, HDR_PARITY);
        byte[][] cws = new byte[nchunks][];
        for (int c = 0; c < nchunks; c++) {
            byte[] block = new byte[k];
            for (int i = 0; i < k; i++) {
                int o = c * k + i;
                if (o < l) { block[i] = msg[o]; }
            }
            cws[c] = rsEncode(block, np);
        }
        byte[] body = interleave(cws);
        byte[] res = new byte[out.length + body.length];
        System.arraycopy(out, 0, res, 0, out.length);
        System.arraycopy(body, 0, res, out.length, body.length);
        return res;
    }

    public static Result decodeMessage(byte[] blob) {
        if (blob.length < HDR_LEN) { throw new RuntimeException("fec: short blob"); }
        byte[] h = rsDecode(Arrays.copyOfRange(blob, 0, HDR_LEN), HDR_PARITY, 5).msg;
        int l = ((h[0] & 0xFF) << 24) | ((h[1] & 0xFF) << 16) | ((h[2] & 0xFF) << 8) | (h[3] & 0xFF);
        int np = h[4] & 0xFF;
        int[] ck = chunking(l, np);
        int nchunks = ck[0];
        int k = ck[1];
        int cwlen = k + np;
        byte[] body = Arrays.copyOfRange(blob, HDR_LEN, blob.length);
        if (body.length != nchunks * cwlen) { throw new RuntimeException("fec: blob length mismatch"); }
        byte[][] cws = deinterleave(body, nchunks, cwlen);
        byte[] out = new byte[nchunks * k];
        int total = 0;
        for (int c = 0; c < nchunks; c++) {
            Result r = rsDecode(cws[c], np, k);
            total += r.corrected;
            System.arraycopy(r.msg, 0, out, c * k, k);
        }
        return new Result(Arrays.copyOfRange(out, 0, l), total);
    }
}
